### ISSUE TITLE
Support for multiple static alertmanagers on the ruler

### ISSUE THREAD

**Issue by gotjosh** [cortexproject/cortex#2860]
Hi everyone 👋,

I've recently opened #2851 as a way to support multiple static alert managers but some concerns were raised about the backwards compatibility of the change. We've had a brief conversation about it [here](https://cloud-native.slack.com/archives/CCYDASBLP/p1594206326383800).

I want to take a step back and get a gauge for which direction we want to take it, the goal of my PR was to introduce the minimal amount of changes to support the feature and I thought a breaking change would be acceptable given the experimental status of the ruler. However, I now realise this might be unfeasible. From here, we have several options.

## Not introduce a breaking change but keep the approach in the PR

I think this is the simplest way to proceed, It would be relatively simple to maintain the existing behaviour and re-use the same `-ruler.alertmanager-url` flag we support today. 

The change here would be that you can now pass multiple URLs in the form of a single string e.g. `-ruler.alertmanager-url=alertmanager1.svc.local,alertmanager2.com` but keep the same behaviour when only 1 URL is supplied. 

There's a bit more to this, with 1 URL choosing things like `username, password, http_prefix, and path prefix` is pretty unambiguous given we can just extract it from the URL itself (that's what we do today); With > 1 not so much. In this case, we can simply document the fact that when you supply more than 1 URL whatever value is defined in the other flags introduced (e.g. `-ruler.alertmanager-scheme`) takes precedence (they all have a default value anyway). 

## Support multiple Alertmanager HA groups

Prometheus [supports multiple groups of Alertmanagers](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#alertmanager_config), under their Scheme you can supply multiple Alertmanager configurations (or groups so to speak) and each group is treated as a HA Alertmanager group. Keep in mind, that each group can include multiple static or discoverable Alertmanagers. It is more of a group of groups philosophy. The key being that alert send failure is claimed only if the Ruler fails to send to all instances of the group.

As it stands today, **Cortex does not follow that philosophy** and is not what my original PR is about. From the Cortex POV, what we're effectively defining is a single group.

If we choose this is the world we want to be in, then this a pretty big change in direction compared to the PR I'm opening. We'd need to discuss whenever we want a mixture of configuration files + flags given we'd now need to define the same additional flags for each group (e.g. HTTP path prefixes for each group). For this particular case, I'd understand a configuration file would work best e.g.

```yaml
ruler:
  alertmanagers:
    - http_path_prefix: ...
       static_config:
         - alertmanager1.com
```

Pretty similar to what [Thanos does](https://thanos.io/components/rule.md/#alertmanager).

I hope this clarifies a bit on what I'm proposing to improve with PR and helps us choose the best way forward for the project.

**Comment by gouthamve** [comment]
I think we should keep the support for `-ruler.alertmanager-url` as is. And add support for multiple AM groups in the config file.

I think we will end up supporting multiple AM groups in the future and the config is not something that can trivially be included in flags.


Finally, for `-ruler.alertmanager-url=alertmanager1.svc.local,alertmanager2.com`, can make each URL part of a different group hence disambiguate the scheme and things? I know that this means if it fails to send to one AM, its a failure, but if you want multiple AMs in a single group, it should be something that is set via config file.

**Comment by gotjosh** [comment]
> Finally, for -ruler.alertmanager-url=alertmanager1.svc.local,alertmanager2.com, can make each URL part of a different group hence disambiguate the scheme and things? I know that this means if it fails to send to one AM, its a failure, but if you want multiple AMs in a single group, it should be something that is set via config file.

That's a good shout. The only downside to this is that we wouldn't really be supporting multiple AMs for a group under static configuration. We'd always need to rely on the DNS service discovery to find the replicas of a group. 

The good thing is that it would eliminate the need for supporting a config file or the other flags in the meantime.

**Comment by pracucci** [comment]
I don't have a strong opinion because I'm not much into the alertmanager. However, the discussion here is not about "how to avoid breaking changes" but more strategic on whether we want to support multiple groups or not (and I don't have a strong opinion on this, so I will commit to whatever decision you will take).

>  I thought a breaking change would be acceptable given the experimental status of the ruler.

Keep in mind the ruler is **not** marked as experimental. Only ruler APIs are.

**Comment by gotjosh** [comment]
Closed by #2851
